aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,231 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NamedTensorUtils.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+
+namespace at {
+namespace native {
+
+// helper
+namespace {
+
+template <typename scalar_t>
+std::vector<int64_t> sort_indices(const TensorAccessor<scalar_t, 1>& v) {
+  std::vector<int64_t> indices(v.size(0));
+  std::iota(indices.begin(), indices.end(), 0);
+  std::stable_sort(
+      indices.begin(),
+      indices.end(),
+      [&](const scalar_t& i1, const scalar_t& i2) { return v[i1] < v[i2]; });
+  return indices;
+}
+
+template <size_t N>
+std::vector<int64_t> indices2csr(
+    const TensorAccessor<int64_t, N>& indices,
+    int64_t dim) {
+  auto nnz = indices.size(0);
+  std::vector<int64_t> csr(dim + 1, int64_t(0));
+  int64_t last_i = 0;
+  for (int64_t index = 0; index < nnz; index++) {
+    int64_t i = indices[index];
+    if (i == last_i) {
+      csr[last_i + 1] += 1;
+    } else {
+      for (int64_t j = last_i; j < i + 1; j++) {
+        csr[j + 1] = csr[last_i + 1];
+      }
+      last_i = i;
+      csr[last_i + 1] += 1;
+    }
+  }
+  for (int64_t j = last_i; j < dim; j++) {
+    csr[j + 1] = csr[last_i + 1];
+  }
+  return csr;
+}
+
+template <typename scalar_t>
+void sparse_matmul_kernel(
+    Tensor& result,
+    const Tensor& self,
+    const Tensor& other) {
+  auto indices_a = self._indices().contiguous();
+  auto values_a = self._values().contiguous();
+
+  auto indices_b = other._indices().contiguous();
+  auto values_b = other._values().contiguous();
+
+  auto nnz_a = values_a.size(0);
+  auto nnz_b = values_b.size(0);
+  auto indices_accessor_a = indices_a.accessor<int64_t, 2>();
+  auto indices_accessor_b = indices_b.accessor<int64_t, 2>();
+
+  auto values_accessor_a = values_a.accessor<scalar_t, 1>();
+  auto values_accessor_b = values_b.accessor<scalar_t, 1>();
+
+  auto result_indices = result._indices().contiguous();
+  auto result_values = result._values().contiguous();
```

Comment:
This is interesting because it might copy the indices and values tensors. However, this PR doesn't expose this out variant to users, right, and it doesn't check that the given result tensor is correct, either... instead of this, should there just be a comment asserting that a condition of calling this function is that the result has certain properties?

Reply:
Yes, It might copy it only if the tensor is not contiguous. For this case, It's redundant as result tensor is allocated in c++ side code.